Make non-alpha/VQE/QAOA grove modules pyQuil2.0 compatible

amyfbrown opened the pull request:
The `alpha` modules have already been migrated and merged into master, and @mpharrigan and I have decided that the `pyvqe`, `pyqaoa`, and `ising` modules will be dealt with in a separate PR because they're heavily interdependent, so this PR migrates the remaining grove modules. This resolves #186.

mpharrigan commented on grove/amplification/grover.py:
```diff
@@ -55,69 +57,72 @@ def _compute_grover_oracle_matrix(bitstring_map):
+        ro = self.grover_circuit.declare('ro', 'BIT', len(self.qubits))
+        self.grover_circuit += [MEASURE(qubit, ro[idx]) for idx, qubit in enumerate(self.qubits)]
+        self.grover_circuit.wrap_in_numshots_loop(1)
+        executable = qc.compiler.native_quil_to_executable(self.grover_circuit)
```
is there a reason why we can't do `executable = qc.compile(self.grover_circuit)`

mpharrigan replied:
👍 just making sure there wasn't some problem with `qc.compile()` :) :)

amyfbrown replied:
Now `qc.run(executable)` gives me a type exception message in PyCharm. It expects a `BinaryExecutableResponse` or a `PyQuilExecutableResponse` but gets a `Message` instead. Sure enough, [`QuantumComputer.compile`](https://github.com/rigetti/pyquil/blob/master/pyquil/api/_quantum_computer.py#L243) has `Message` as a return type. 

Does this change your mind about replacing `qc.compiler.native_quil_to_executable` with `qc.compile`, @mpharrigan?

mpharrigan commented on grove/bernstein_vazirani/bernstein_vazirani.py:
```diff
@@ -101,10 +102,10 @@ def _compute_unitary_oracle_matrix(bitstring_map):
 
         :param Dict[String, String] bitstring_map: truth-table of the input bitstring map in
-        dictionary format
+            dictionary format
```
👍

mpharrigan commented on grove/bernstein_vazirani/bernstein_vazirani.py:
```diff
@@ -179,17 +178,30 @@ def run(self, cxn, bitstring_map):
+        full_circuit += [MEASURE(qubit, ro) for qubit, ro in zip(self.computational_qubits,
+                                                                 full_ro)]
+        full_circuit.wrap_in_numshots_loop(1)
+        full_executable = qc.compiler.native_quil_to_executable(full_circuit)
```
same q as above

mpharrigan commented on grove/bernstein_vazirani/bernstein_vazirani.py:
```diff
@@ -179,17 +178,30 @@ def run(self, cxn, bitstring_map):
+        full_circuit.wrap_in_numshots_loop(1)
+        full_executable = qc.compiler.native_quil_to_executable(full_circuit)
+        full_results = qc.run(full_executable)
+        bv_vector = full_results[0][::-1]
```
worth measuring into classical addresses indexed in a descending order so you don't have to flip the bitstring?

mpharrigan replied:
maybe not worth it

msohaibalam commented on grove/bernstein_vazirani/bernstein_vazirani.py:
```diff
@@ -179,17 +178,30 @@ def run(self, cxn, bitstring_map):
+        full_circuit += self.bv_circuit
+        full_circuit += [MEASURE(qubit, ro) for qubit, ro in zip(self.computational_qubits,
+                                                                 full_ro)]
+        full_circuit.wrap_in_numshots_loop(1)
```
You don't need this line at all if you're only doing a single shot. You may have to change the indexing in `full_results` to obtain `bv_vector` below though if you do remove this line.

msohaibalam replied:
Correcting myself, you probably *don't* need to change the indexing of `full_results` below, if you remove this line, since `Program().wrap_in_numshots_loop(1)` seems to happen by default anyway.

msohaibalam commented on grove/bernstein_vazirani/bernstein_vazirani.py:
```diff
@@ -179,17 +178,30 @@ def run(self, cxn, bitstring_map):
+        ancilla_ro = ancilla_circuit.declare('ro', 'BIT', len(self.computational_qubits) + 1)
+        ancilla_circuit += self.bv_circuit
+        ancilla_circuit += [MEASURE(self.ancilla, ancilla_ro[self.ancilla])]
+        ancilla_circuit.wrap_in_numshots_loop(1)
```
Same comment as above.

msohaibalam commented on grove/deutsch_jozsa/deutsch_jozsa.py:
```diff
@@ -53,13 +56,20 @@ def is_constant(self, cxn, bitstring_map):
+        dj_ro = prog.declare('ro', 'BIT', len(self.computational_qubits))
+        prog += self.deutsch_jozsa_circuit
+        prog += [MEASURE(qubit, ro) for qubit, ro in zip(self.computational_qubits, dj_ro)]
+        prog.wrap_in_numshots_loop(1)
```
Same as previous comment about not needing this for single shot results.

msohaibalam commented on grove/simon/simon.py:
```diff
@@ -246,22 +242,28 @@ def find_mask(self, cxn, bitstring_map):
+            simon_ro = prog.declare('ro', 'BIT', len(self.computational_qubits))
+            prog += self.simon_circuit
+            prog += [MEASURE(qubit, ro) for qubit, ro in zip(self.computational_qubits, simon_ro)]
+            prog.wrap_in_numshots_loop(1)
```
Same as comment(s) above

amyfbrown commented:
@msohaibalam, @mpharrigan, thanks for the review! I changed `qc.compiler.native_quil_to_executable` to `qc.compile` and removed `wrap_in_numshots_loop(1)`. Can I bug you for another look at your earliest convenience? 😃
